File: src/adapters/php/src/LpdfKit.php
<?php

declare(strict_types=1);

namespace Lpdf;

/**
 * Static tree-builder helpers for constructing lpdf document trees programmatically.
 *
 * All helpers return plain serialisable objects. Pass the result of
 * {@see document()} to {@see LpdfEngine::renderPdf()}.
 *
 * @example
 * ```php
 * use Lpdf\{LpdfKit, LpdfEngine, PageOptions, DocumentOptions, LpdfMeta};
 *
 * $doc = LpdfKit::document(
 *     nodes: [
 *         LpdfKit::page(
 *             nodes: [LpdfKit::text(nodes: ['Hello, world!'])],
 *             options: new PageOptions(size: 'a4', margin: '28pt'),
 *         ),
 *     ],
 *     options: new DocumentOptions(meta: new LpdfMeta(title: 'My Doc')),
 * );
 * $pdf = (new LpdfEngine(''))->renderPdf($doc);
 * ```
 */
final class LpdfKit
{
    // ── Container helpers ─────────────────────────────────────────────────────

    /** @param LpdfNode[] $nodes */
    public static function stack(array $nodes = [], ?StackOptions $options = null): LpdfContainerNode
    {
        return new LpdfContainerNode('stack', self::optionsToAttrs($options), $nodes);
    }

    /** @param LpdfNode[] $nodes */
    public static function flank(array $nodes = [], ?FlankOptions $options = null): LpdfContainerNode
    {
        return new LpdfContainerNode('flank', self::optionsToAttrs($options), $nodes);
    }

    /** @param LpdfNode[] $nodes */
    public static function split(array $nodes = [], ?SplitOptions $options = null): LpdfContainerNode
    {
        return new LpdfContainerNode('split', self::optionsToAttrs($options), $nodes);
    }

    /** @param LpdfNode[] $nodes */
    public static function cluster(array $nodes = [], ?ClusterOptions $options = null): LpdfContainerNode
    {
        return new LpdfContainerNode('cluster', self::optionsToAttrs($options), $nodes);
    }

    /** @param LpdfNode[] $nodes */
    public static function grid(array $nodes = [], ?GridOptions $options = null): LpdfContainerNode
    {
        return new LpdfContainerNode('grid', self::optionsToAttrs($options), $nodes);
    }

    /** @param LpdfNode[] $nodes */
    public static function frame(array $nodes = [], ?FrameOptions $options = null): LpdfContainerNode
    {
        return new LpdfContainerNode('frame', self::optionsToAttrs($options), $nodes);
    }

    /** @param LpdfNode[] $nodes */
    public static function link(array $nodes = [], ?LinkOptions $options = null): LpdfContainerNode
    {
        return new LpdfContainerNode('link', self::optionsToAttrs($options), $nodes);
    }

    // ── Leaf helpers ──────────────────────────────────────────────────────────

    /**
     * Build a text paragraph node.
     *
     * @param  array<string|LpdfSpanNode> $nodes Children must be strings or LpdfSpanNode instances.
     * @throws \InvalidArgumentException         if a child is neither a string nor a LpdfSpanNode.
     */
    public static function text(array $nodes = [], ?TextOptions $options = null): LpdfTextNode
    {
        foreach ($nodes as $i => $child) {
            if (!is_string($child) && !$child instanceof LpdfSpanNode) {
                throw new \InvalidArgumentException(
                    "text() child at index $i must be a string or LpdfSpanNode, got " . get_debug_type($child),
                );
            }
        }

        return new LpdfTextNode(self::optionsToAttrs($options), $nodes);
    }

    /**
     * Build a span inline node.
     *
     * @param  string[] $nodes Children must be plain strings.
     * @throws \InvalidArgumentException if a child is not a string.
     */
    public static function span(array $nodes = [], ?SpanOptions $options = null): LpdfSpanNode
    {
        foreach ($nodes as $i => $child) {
            if (!is_string($child)) {
                throw new \InvalidArgumentException(
                    "span() child at index $i must be a string, got " . get_debug_type($child),
                );
            }
        }

        return new LpdfSpanNode(self::optionsToAttrs($options), $nodes);
    }

    /** Build a divider (horizontal rule) node. */
    public static function divider(?DividerOptions $options = null): LpdfDividerNode
    {
        return new LpdfDividerNode(self::optionsToAttrs($options));
    }

    // ── Page + document ───────────────────────────────────────────────────────

    /** @param LpdfNode[] $nodes */
    public static function page(array $nodes = [], ?PageOptions $options = null): LpdfPageNode
    {
        return new LpdfPageNode(self::optionsToAttrs($options), $nodes);
    }

    /**
     * Build the root document node, ready for {@see LpdfEngine::renderPdf()}.
     *
     * @param LpdfPageNode[] $nodes
     */
    public static function document(array $nodes = [], ?DocumentOptions $options = null): LpdfDocument
    {
        // tokens and meta are sub-objects, not flat string attrs — handle separately.
        $attrs = self::optionsToAttrs($options, skip: ['tokens', 'meta']);

        if ($options?->tokens !== null) {
            $attrs['tokens'] = $options->tokens;
        }
        if ($options?->meta !== null) {
            $attrs['meta'] = $options->meta;
        }

        return new LpdfDocument($attrs, $nodes);
    }

    // ── Private helpers ───────────────────────────────────────────────────────

    /**
     * Reflect over a readonly options object and convert its non-null string
     * properties to a kebab-case attribute map, skipping named properties.
     *
     * @param  string[] $skip Property names to exclude (handled separately).
     * @return array<string,string>
     */
    private static function optionsToAttrs(?object $options, array $skip = []): array
    {
        if ($options === null) {
            return [];
        }

        $attrs = [];
        foreach ((new \ReflectionClass($options))->getProperties() as $prop) {
            $name = $prop->getName();
            if (in_array($name, $skip, true)) {
                continue;
            }
            $value = $prop->getValue($options);
            if (!is_string($value)) {
                continue;
            }
            $attrs[self::camelToKebab($name)] = $value;
        }

        return $attrs;
    }

    /** camelCase / PascalCase → kebab-case. */
    private static function camelToKebab(string $name): string
    {
        return strtolower((string) preg_replace('/[A-Z]/', '-$0', lcfirst($name)));
    }
}

File: src/adapters/dotnet/src/LpdfKit.cs
using System.Collections.Concurrent;
using System.Reflection;

namespace Lpdf;

/// <summary>
/// Static tree-builder helpers for constructing lpdf document trees.
///
/// All helpers return plain serialisable records. Pass the result of
/// <see cref="Document"/> to <see cref="LpdfEngine.RenderPdf(LpdfDocument, RenderOptions?)"/>.
///
/// <example>
/// <code>
/// using Lpdf;
///
/// var engine = new LpdfEngine(licenseKey);
/// var bytes  = await engine.RenderPdf(
///     LpdfKit.Document(new DocumentInput {
///         Nodes = [
///             LpdfKit.Page(new PageInput {
///                 Nodes = [LpdfKit.Text(new TextInput { Nodes = ["Hello, world!"] })],
///                 Options = new PageOptions { Size = "a4", Margin = "28pt" },
///             }),
///         ],
///         Options = new DocumentOptions {
///             Meta = new LpdfMeta(Title: "My Doc"),
///         },
///     }));
/// </code>
/// </example>
/// </summary>
public static class LpdfKit
{
    // ── Container helpers ─────────────────────────────────────────────────────

    /// <summary>Build a <c>stack</c> layout node (vertical column).</summary>
    public static LpdfNode Stack(StackInput input) =>
        Container("stack", input.Options, input.Nodes);

    /// <summary>Build a <c>flank</c> layout node (horizontal row).</summary>
    public static LpdfNode Flank(FlankInput input) =>
        Container("flank", input.Options, input.Nodes);

    /// <summary>Build a <c>split</c> layout node (two-column split).</summary>
    public static LpdfNode Split(SplitInput input) =>
        Container("split", input.Options, input.Nodes);

    /// <summary>Build a <c>cluster</c> layout node (wrapping flex row).</summary>
    public static LpdfNode Cluster(ClusterInput input) =>
        Container("cluster", input.Options, input.Nodes);

    /// <summary>Build a <c>grid</c> layout node (multi-column grid).</summary>
    public static LpdfNode Grid(GridInput input) =>
        Container("grid", input.Options, input.Nodes);

    /// <summary>Build a <c>frame</c> layout node (fixed-size container).</summary>
    public static LpdfNode Frame(FrameInput input) =>
        Container("frame", input.Options, input.Nodes);

    /// <summary>Build a <c>link</c> layout node (hyperlink wrapper).</summary>
    public static LpdfNode Link(LinkInput input) =>
        Container("link", input.Options, input.Nodes);

    // ── Leaf helpers ──────────────────────────────────────────────────────────

    /// <summary>Build a <c>divider</c> horizontal rule node.</summary>
    public static LpdfDividerNode Divider(DividerInput input) =>
        new(Attrs(input.Options));

    /// <summary>Build an <c>img</c> image node.</summary>
    public static LpdfImgNode Img(ImgInput input) =>
        new(Attrs(input.Options));

    /// <summary>Build a <c>barcode</c> node.</summary>
    public static LpdfBarcodeNode Barcode(BarcodeInput input) =>
        new(Attrs(input.Options));

    // ── Table helpers ─────────────────────────────────────────────────────────

    /// <summary>Build a <c>table</c> layout node.</summary>
    public static LpdfContainerNode Table(TableInput input) =>
        Container("table", input.Options, input.Nodes);

    /// <summary>Build a <c>thead</c> table header row group.</summary>
    public static LpdfContainerNode Thead(TheadInput input) =>
        Container("thead", input.Options, input.Nodes);

    /// <summary>Build a <c>tr</c> table row.</summary>
    public static LpdfContainerNode Tr(TrInput input) =>
        Container("tr", input.Options, input.Nodes);

    /// <summary>Build a <c>td</c> table cell.</summary>
    public static LpdfContainerNode Td(TdInput input) =>
        Container("td", input.Options, input.Nodes);

    // ── Text helpers ──────────────────────────────────────────────────────────

    /// <summary>Build a <c>text</c> paragraph node.</summary>
    public static LpdfTextNode Text(TextInput input) => new(
        Attrs(input.Options),
        (input.Nodes ?? []).ToList());

    /// <summary>Build a <c>span</c> inline node.</summary>
    public static LpdfSpanNode Span(SpanInput input) => new(
        Attrs(input.Options),
        (input.Nodes ?? []).ToList());

    // ── Page + document ───────────────────────────────────────────────────────

    /// <summary>Build a <c>page</c> node.</summary>
    public static LpdfPageNode Page(PageInput input) => new(
        Attrs(input.Options),
        (input.Nodes ?? []).ToList());

    /// <summary>Build the root <c>document</c> node, ready for <see cref="LpdfEngine.RenderPdf(LpdfDocument, RenderOptions?)"/>.</summary>
    public static LpdfDocument Document(DocumentInput input)
    {
        var opts   = input.Options ?? new DocumentOptions();
        var attrs  = new Dictionary<string, object?>(StringComparer.Ordinal);

        // Flat string attrs
        if (opts.Size        is not null) attrs["size"]        = opts.Size;
        if (opts.Orientation is not null) attrs["orientation"] = opts.Orientation;
        if (opts.Margin      is not null) attrs["margin"]      = opts.Margin;
        if (opts.Background  is not null) attrs["background"]  = opts.Background;

        // Tokens sub-object
        if (opts.Tokens is not null)
            attrs["tokens"] = SerializeTokens(opts.Tokens);

        // Meta sub-object
        if (opts.Meta is not null)
            attrs["meta"] = SerializeMeta(opts.Meta);

        return new LpdfDocument(attrs, (input.Nodes ?? []).ToList());
    }

    // ── Implicit conversion helper ────────────────────────────────────────────

    /// <summary>
    /// Wrap a plain string as <see cref="LpdfContent"/> for use in
    /// <see cref="TextInput.Nodes"/>.
    /// </summary>
    public static LpdfContent Text(string raw) => new LpdfRawText(raw);

    // ── Private helpers ───────────────────────────────────────────────────────

    private static LpdfContainerNode Container(string type, object? options, LpdfNode[]? children) =>
        new(Attrs(options), (children ?? []).ToList(), type);

    /// <summary>
    /// Reflect over any options record and convert its non-null properties to
    /// a kebab-case attribute dictionary.
    /// </summary>
    private static readonly ConcurrentDictionary<Type, PropertyInfo[]> _propCache = new();

    private static Dictionary<string, string> Attrs(object? options)
    {
        var result = new Dictionary<string, string>(StringComparer.Ordinal);
        if (options is null) return result;

        var props = _propCache.GetOrAdd(
            options.GetType(),
            t => t.GetProperties(BindingFlags.Public | BindingFlags.Instance));
        foreach (var prop in props)
        {
            var value = prop.GetValue(options);
            if (value is string s)
                result[PascalToKebab(prop.Name)] = s;
        }
        return result;
    }

    private static Dictionary<string, object?> SerializeTokens(LpdfTokens t)
    {
        var d = new Dictionary<string, object?>(StringComparer.Ordinal);
        if (t.Colors is not null) d["colors"] = t.Colors;
        if (t.Space  is not null) d["space"]  = t.Space;
        if (t.Grid   is not null) d["grid"]   = t.Grid;
        if (t.Border is not null) d["border"] = t.Border;
        if (t.Radius is not null) d["radius"] = t.Radius;
        if (t.Width  is not null) d["width"]  = t.Width;
        if (t.Text   is not null) d["text"]   = t.Text;
        if (t.Fonts  is not null)
        {
            var fonts = new Dictionary<string, object?>(StringComparer.Ordinal);
            foreach (var (name, def) in t.Fonts)
                fonts[name] = def switch
                {
                    LpdfFontSrc     src     => (object)new { src     = src.Src },
                    LpdfFontBuiltin builtin => (object)new { builtin = builtin.Builtin },
                    _                       => null,
                };
            d["fonts"] = fonts;
        }
        return d;
    }

    private static Dictionary<string, string?> SerializeMeta(LpdfMeta m) =>
        new(StringComparer.Ordinal)
        {
            ["title"]    = m.Title,
            ["author"]   = m.Author,
            ["subject"]  = m.Subject,
            ["keywords"] = m.Keywords,
            ["creator"]  = m.Creator,
        };

    /// <summary>PascalCase → kebab-case (e.g. <c>FontSize</c> → <c>font-size</c>).</summary>
    private static string PascalToKebab(string name)
    {
        var sb = new System.Text.StringBuilder(name.Length + 4);
        for (var i = 0; i < name.Length; i++)
        {
            var c = name[i];
            if (char.IsUpper(c) && i > 0) sb.Append('-');
            sb.Append(char.ToLowerInvariant(c));
        }
        return sb.ToString();
    }
}
